src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,608 @@
+            {
+                if (_semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any() ||
+                    source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                        source.ContainsDirectives))
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                return TryConvertInternal(source, source.Parent, out documentUpdate);
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, SyntaxNode parent, out DocumentUpdate documentUpdate)
+            {
+                switch (parent.Kind())
+                {
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertIfInReturnStatement(source, (ReturnStatementSyntax)parent, out documentUpdate);
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertIfInForEach(source, (ForEachStatementSyntax)parent, out documentUpdate);
+                    case SyntaxKind.EqualsValueClause:
+                    case SyntaxKind.SimpleAssignmentExpression:
+                        return TryReplaceWithLocalFunction(source, parent, out documentUpdate);
+                    case SyntaxKind.ParenthesizedExpression:
+                        return TryConvertInternal(source, parent.Parent, out documentUpdate);
+                    case SyntaxKind.SimpleMemberAccessExpression:
+                        return TryConvertIfInMemberAccessExpression(source, (MemberAccessExpressionSyntax)parent, out documentUpdate);
+                    default:
+                        documentUpdate = default;
+                        return false;
+                }
+            }
+
+            private bool TryConvertIfInMemberAccessExpression(
+               QueryExpressionSyntax source,
+               MemberAccessExpressionSyntax memberAccessExpression,
+               out DocumentUpdate documentUpdate)
+            {
+                if (memberAccessExpression.Parent is InvocationExpressionSyntax invocationExpression)
+                {
+                    switch (memberAccessExpression.Name.Identifier.ValueText)
+                    {
+                        case nameof(Enumerable.ToList):
+                            return TryConvertIfInToListInvocation(source, invocationExpression, out documentUpdate);
+                        case nameof(Enumerable.Count):
+                            return TryConvertIfInCountInvocation(source, invocationExpression, out documentUpdate);
+                        default:
+                            return TryReplaceWithLocalFunction(source, memberAccessExpression, out documentUpdate);
```

Comment:
i find some of the logic harder to follow without example cases.  Can you give comments with simple examples of what sort of code would call through to here, and what we want to convert it to.